README.md
# hw.rest-parser

Practice assignment to implement a command-line app that takes an input file & outputs sorted views.

As this isn't a "real production app", this repo's code has some metacommentary about _why_ particular decisions were made.

There are two phases of this app's development:

1. A simple CLI app that takes an input file & outputs three different views of the sorted records
   * Automatically recognizes input file format
   * Renders the three sorted views in order
   * Expects input files to be consistently in one of the three described formats
   * code at https://github.com/Quezion/hw.rest-parser/releases/tag/Step-1

2. A rest HTTP server that receives new records & return sorted variations of them on endpoints

## Usage

Launch a REPL session with your editor of choice (i.e. `M-x cider-jack-in-clj` or `lein repl`)

### CLI

See comment at end of `src/hw/rest_parser.clj` for REPL usage examples.

Run directly via `lein` like below.

```bash
export FILEPATHS="data/basic_record.csv,data/basic_record.psv,data/basic_record.ssv"
lein run
```

You may optionally set the `column-length` to be used when rendering the output table.
```
export COLUMN_LENGTH=10
```

## Test

Run at REPL or use `lein test`

## License

Copyright © 2021 Quest Yarbrough

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-10-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-10-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/hw.rest-parser/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/hw.rest-parser/compare/0.1.0...0.1.1


test/hw/rest_parser_test.clj
(ns hw.rest-parser-test
  (:require [environ.core :as environ]
            [clojure.edn :as edn]
            [clojure.string :as str]
            [clojure.test :refer :all]
            [hw.rest-parser :refer :all]
            [malli.instrument :as mi]))

;; We'd normally handle this via state-management lib to allow instrumentation in any app flow
;; Not only in tests, but also in any CI environments. (How confident are you in your specs? :)
(mi/instrument!)

(def expected-table
  "LastName   FirstName  Email      Favorit... DateOfB...
  ======================================================
  Pierce     Dennis     dennis.... black      9/7/1984
  Fitzpat... Donna      donna@f... blue       8/10/1972
  Mcclure    Joey       joey.mc... magenta    3/22/1991
  Morris     Tyler      tyler.m... orange     8/6/1951
  Stewart    Dora       dora.st... purple     9/6/1976
  Price      Lucille    lucille... silver     1/1/1946

  LastName   FirstName  Email      Favorit... DateOfB...
  ======================================================
  Price      Lucille    lucille... silver     1/1/1946
  Morris     Tyler      tyler.m... orange     8/6/1951
  Fitzpat... Donna      donna@f... blue       8/10/1972
  Stewart    Dora       dora.st... purple     9/6/1976
  Pierce     Dennis     dennis.... black      9/7/1984
  Mcclure    Joey       joey.mc... magenta    3/22/1991

  LastName   FirstName  Email      Favorit... DateOfB...
  ======================================================
  Stewart    Dora       dora.st... purple     9/6/1976
  Price      Lucille    lucille... silver     1/1/1946
  Pierce     Dennis     dennis.... black      9/7/1984
  Morris     Tyler      tyler.m... orange     8/6/1951
  Mcclure    Joey       joey.mc... magenta    3/22/1991
  Fitzpat... Donna      donna@f... blue       8/10/1972")

(defn trim-lines [table]
  (->> (str/split table #"\n")
       (map str/trim)
       (str/join "\n")))

(deftest main-test
  (testing "Merged file records return expected string table"
    (with-env {:filepaths "data/basic_record.csv,data/basic_record.psv,data/basic_record.ssv"
               :column-length 10}
      (is (= (trim-lines (with-out-str (-main)))
             ;; https://www.youtube.com/watch?v=QyrDgEz3DR0
             (trim-lines expected-table))))))

(deftest run-test
  (testing "Basic CSV record loading returns expected results"
    (is (= (run {:filepaths "data/basic_record.csv"})
           '({:LastName "Mcclure" :FirstName "Joey" :Email "joey.mc@gmail.com"
              :FavoriteColor "magenta" :DateOfBirth #inst "1991-03-22T08:00:00.000-00:00"}
             {:LastName "Fitzpatrick" :FirstName "Donna" :Email "donna@fitzpatrick.com"
              :FavoriteColor "blue" :DateOfBirth #inst "1972-08-10T07:00:00.000-00:00"}))))

  (testing "Merged file records return expected results"
    (is (= (run {:filepaths "data/basic_record.csv,data/basic_record.psv,data/basic_record.ssv"})
           '({:LastName "Mcclure"
              :FirstName "Joey"
              :Email "joey.mc@gmail.com"
              :FavoriteColor "magenta"
              :DateOfBirth #inst "1991-03-22T08:00:00.000-00:00"}
             {:LastName "Fitzpatrick"
              :FirstName "Donna"
              :Email "donna@fitzpatrick.com"
              :FavoriteColor "blue"
              :DateOfBirth #inst "1972-08-10T07:00:00.000-00:00"}
             {:LastName "Morris"
              :FirstName "Tyler"
              :Email "tyler.morris@example.com"
              :FavoriteColor "orange"
              :DateOfBirth #inst "1951-08-06T07:00:00.000-00:00"}
             {:LastName "Stewart"
              :FirstName "Dora"
              :Email "dora.stewart@example.com"
              :FavoriteColor "purple"
              :DateOfBirth #inst "1976-09-06T07:00:00.000-00:00"}
             {:LastName "Pierce"
              :FirstName "Dennis"
              :Email "dennis.pierce@example.com"
              :FavoriteColor "black"
              :DateOfBirth #inst "1984-09-07T07:00:00.000-00:00"}
             {:LastName "Price"
              :FirstName "Lucille"
              :Email "lucille.price@example.com"
              :FavoriteColor "silver"
              :DateOfBirth #inst "1946-01-01T08:00:00.000-00:00"})))))

(def minimal-unsorted-records
  "Small handrolled set to guarantee correct sorting"
  [{:FavoriteColor "blue" :LastName "Parker"
    :DateOfBirth #inst "1991-03-22T08:00:00.000-00:00"}
   {:FavoriteColor "red" :LastName "Bombadil"
    :DateOfBirth #inst "1972-08-10T07:00:00.000-00:00"}
   {:FavoriteColor "blue" :LastName "Packer"
    :DateOfBirth #inst "1951-08-06T07:00:00.000-00:00"}
   {:FavoriteColor "pink" :LastName "Stelsior"
    :DateOfBirth #inst "1984-09-07T07:00:00.000-00:00"}])

(deftest sort-records-test
  (let [[parker bombadil packer stelsior] minimal-unsorted-records]
    (is (= (sort-records minimal-unsorted-records lastname-desc)
           [stelsior parker packer bombadil]))
    (is (= (sort-records minimal-unsorted-records dob)
           [packer bombadil stelsior parker]))
    (is (= (sort-records minimal-unsorted-records color-lastname)
           [packer parker stelsior bombadil]))))


test/hw/rest_parser/server-test.clj
(ns hw.rest-parser.cli-test
  (:require [environ.core :as environ]
            [clojure.edn :as edn]
            [clojure.string :as str]
            [clojure.test :refer :all]
            [hw.rest-parser.server :refer :all]
            [malli.instrument :as mi]))


project.clj
(defproject hw.rest-parser "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.3"]

                 ;; Provides map of instantiated $environment vars at environ.core/env
                 ;; https://github.com/weavejester/environ
                 [environ "1.2.0"]

                 ;; Wraps clojure.spec to validate data structures & provide FN instrumentation.
                 ;; Cleaner than base spec with extra features & more runtime dynamicism.
                 [metosin/malli "0.6.2"]

                 ;; Simple state management library. There's good arguments for others,
                 ;; but Mount is sufficient for this simple repo.
                 [mount "0.1.16"]

                 ;; Modern HTTP client + server. Usually picked for its option of full server asynchronicity.
                 ;; Battle-proven in prod at several millions of requests / day.
                 [aleph "0.4.6"]]
  :plugins [[lein-environ "1.2.0"]]
  :main ^:skip-aot hw.rest-parser.server
  :target-path "target/%s"
  :profiles {:test {:resource-paths ["data"]}
             :uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}})


src/hw/rest_parser.clj
(ns hw.rest-parser
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [environ.core :refer [env]]
            [malli.core :as m]
            [malli.error :as me])
  (:import [java.text SimpleDateFormat])
  (:gen-class))

(def row-headers [:LastName :FirstName :Email :FavoriteColor :DateOfBirth])

(defn line-match-separator
  "Returns regex for separator of string line, or nil if line doesn't have consistent separator.
  separators - set of string separators that could be between values
  line - string with "
  [separators line]
  (->> (map #(when (= (count (re-seq % line))
                      (- (count row-headers) 1))
               %)
            separators)
       (filter some?)
       first))

(def allowed-separators-regexp
  "Allowed line separators in correct sequence for line-matching"
  [#","
   #"\|"
   #" "])

(def match-allowed-separator
  "Matches line against all allowed separators & return first match"
  (partial line-match-separator allowed-separators-regexp))

(defn line->separator
  "Returns one of known valid string separators or nil if none could be determined"
  [line]
  (match-allowed-separator line))

(defn parse-datestring
  "Parses date in expected format of MM/DD/YYYY"
  [date]
  (-> (SimpleDateFormat. "MM/dd/yyyy")
      (.parse date)))

(defn load-record
  "Loads record at target filepath into memory.
  A more graceful implementation might hold files open for the entire runtime,
  but side-effect complexity becomes messy without care & is better skipped for a sample app"
  [filepath]
  ;; Assume no header lines at top of input files
  (with-open [rdr (io/reader filepath)]
    (let [lines (line-seq rdr)
          separator (some-> (first lines)
                            (line->separator))]
      (when separator
        (->> lines
             (map (comp #(update % :DateOfBirth parse-datestring)
                        #(zipmap row-headers %)
                        #(map str/trim %)
                        #(str/split % separator)))
             doall))))) ;; Force in-memory realization now before file closes

(def env-spec [:map [:filepaths [:string {:min 1}]
                     :column-length [:string {:optional true
                                              :min 1}]]])

(defn validate-env [] (m/validate env-spec env))
(defn explain-env [] (m/explain env-spec env))

(defn string-compare-lowercase-asc [a b] (compare (str/lower-case a) (str/lower-case b)))
(defn string-compare-lowercase-desc [a b] (compare (str/lower-case b) (str/lower-case a)))

(def color-lastname [[:FavoriteColor string-compare-lowercase-asc]
                     [:LastName string-compare-lowercase-asc]])

(def dob [[:DateOfBirth compare]])

(def lastname-desc [[:LastName string-compare-lowercase-desc]])

(defn sort-records
  "Returns sorted list of input records based on sort-definition.
  records - seq of maps representing parsed records
  sort-definition - vec of vecs, where each 2-element subvec is keyfn & comparator"
  [records sort-definition]
  (let [sort-keys (map first sort-definition)
        keyfn (apply juxt sort-keys)
        sort-comparators (map second sort-definition)
        comparator (fn [a b]
                     (loop [[comparator & comp-rest] sort-comparators
                            [a1 & a-rest] a
                            [b1 & b-rest] b]
                       (let [result (comparator a b)]
                         (if (and (zero? result)
                                  (seq sort-comparators))
                           (recur comp-rest a-rest b-rest)
                           result))))]
    (sort-by keyfn comparator records)))

(defn run
  [{:keys [filepaths]}]
  (as-> filepaths <>
    (str/split filepaths #",")
    (map load-record <>)
    (flatten <>)))

(def record-spec [:map [:LastName [:string {:min 1}]
                        :FirstName [:string {:min 1}]
                        :Email [:string {:min 5}]
                        :FavoriteColor [:string {:min 4 :max 40}]
                        :DateOfBirth inst?]])

;; Register `run` with Malli. This definition reads "takes env-spec & returns record"
(m/=> run
      [:=> [:cat env-spec]
       [:sequential record-spec]])

(def output-views [color-lastname
                   dob
                   lastname-desc])

(defn trunc
  "This is unexpectedly missing in clojure.string"
  [s n]
  (subs s 0 (min (count s) n)))

(defn force-str-to-length
  "Forces string to specific width, padding with empty spaces as necessary.
  Longer strings are truncated with trailing ..."
  [length s]
  (let [difference (- length (count s))]
    (cond
      (pos? difference) (apply str s (repeat difference " "))
      (neg? difference) (str (trunc s (- length 3)) "...")
      :else s)))

(defn render-view-cast-to-string
  "Casts any record value to string"
  [x]
  (cond
    (string? x) x
    (inst? x) (.format (SimpleDateFormat. "M/d/yyyy") x)
    :else (throw (Exception. (str "Invalid value <x,typeOf(x)> " x (type x))))))

;; Little kludge to derive this from env. In a larger project,
;; we'd parse env into a final map, validate it, & expose it via state management lib
(defn env-column-length
  []
  (try
    (if (integer? (:column-length environ.core/env))
      (:column-length environ.core/env)
      (Integer/parseInt (:column-length environ.core/env)))
    (catch Exception _
      16)))

(defn render-view-data
  "Given sequence of maps with common keys, renders into ASCII table of minimum width"
  ([mapseq]
   (render-view-data mapseq nil))
  ([mapseq {:keys [column-length]
            :or {column-length (env-column-length)}}]
   (let [map-keys (keys (first mapseq))
         labels (->> (map name map-keys)
                     (map #(-> (force-str-to-length column-length %)
                               (str " ")))
                     (str/join)
                     (drop-last)
                     (str/join ""))
         lines (->> (map vals mapseq)
                    (map #(map (fn [s]
                                 (str (->> (render-view-cast-to-string s)
                                           (force-str-to-length column-length))
                                      " ")) %))
                    (map (comp str/join drop-last str/join))
                    (str/join "\n"))]
     (str labels "\n"
          (str/join (repeat (count labels) "="))
          "\n"
          lines))))

(defn run-render-views
  [env]
  (as-> (run (merge environ.core/env env)) <>
    (repeat <>)
    (map sort-records <> output-views)
    (map render-view-data <>)))

(defn -main
  [& _]
  ;; You could accept a file via Java program ARGS... but for sake of simplicity, just use $env
  (if-not (validate-env)
    (->> (explain-env)
         (me/humanize)
         (prn-str)
         (str "[ERROR] Problems below with input $env variables. Please fix & retry\n")
         (println))
    (->> (run-render-views env)
         (str/join "\n\n")
         (println))))

;; This technically belongs in user.clj or test, but including here to make usage obvious
(defmacro with-env
  [env & body]
  `(with-redefs [environ.core/env ~env]
     ~@body))

(comment ;; Try at a REPL. Modify to input different files
  (with-env {:filepaths "data/basic_record.csv,data/basic_record.psv,data/basic_record.ssv"
             :column-length 12}
    (-main)))


src/hw/rest_parser/server.clj
(ns hw.rest-parser.server
  (:require [aleph.http :as http]
            [byte-streams :as bs]
            [clojure.string :as str]
            [environ.core :refer [env]]
            [malli.core :as m]
            [malli.error :as me]
            [mount.core :refer [defstate] :as mount])
  (:import [java.text SimpleDateFormat])
  (:gen-class))

(def defaults {:host "localhost"
               :port 3000})

(defstate webserver-host
  :start (if (seq (:webserver-host env))
           (:webserver-host env)
           (:host defaults)))

(defstate webserver-port
  :start (try (Integer/parseInt (str (:webserver-port env)))
              (catch Exception _
                (:port defaults))))

(defn handler [req]
  {:status 200
   :headers {"content-type" "text/plain"}
   :body "hello!"})

(defstate server
  :start (do (println "[WEB] Starting server on host:port"
                      (str webserver-host ":" webserver-port))
             (http/start-server handler {:host webserver-host
                                         :port webserver-port}))
  :stop (.close server))

(def env-spec [:map
               [:webserver-host {:min 1
                                 :optional true} :string]
               [:webserver-port {:min 1
                                 :optional true} :string]])

(defn validate-env [] (m/validate env-spec env))
(defn explain-env [] (m/explain env-spec env))

(defn -main
  [& _]
  (if-not (validate-env)
    (->> (explain-env)
         (me/humanize)
         (prn-str)
         (str "[ERROR] Problems below with input $env variables. Please fix & retry\n")
         (println))
    (do (mount/start)
        (println "[SUCCESS] App initialized"))))


src/hw/rest_parser/cli.clj
(ns hw.rest-parser.cli
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [environ.core :refer [env]]
            [malli.core :as m]
            [malli.error :as me])
  (:import [java.text SimpleDateFormat])
  (:gen-class))

(def row-headers [:LastName :FirstName :Email :FavoriteColor :DateOfBirth])

(defn line-match-separator
  "Returns regex for separator of string line, or nil if line doesn't have consistent separator.
  separators - set of string separators that could be between values
  line - string with "
  [separators line]
  (->> (map #(when (= (count (re-seq % line))
                      (- (count row-headers) 1))
               %)
            separators)
       (filter some?)
       first))

(def allowed-separators-regexp
  "Allowed line separators in correct sequence for line-matching"
  [#","
   #"\|"
   #" "])

(def match-allowed-separator
  "Matches line against all allowed separators & return first match"
  (partial line-match-separator allowed-separators-regexp))

(defn line->separator
  "Returns one of known valid string separators or nil if none could be determined"
  [line]
  (match-allowed-separator line))

(defn parse-datestring
  "Parses date in expected format of MM/DD/YYYY"
  [date]
  (-> (SimpleDateFormat. "MM/dd/yyyy")
      (.parse date)))

(defn load-record
  "Loads record at target filepath into memory.
  A more graceful implementation might hold files open for the entire runtime,
  but side-effect complexity becomes messy without care & is better skipped for a sample app"
  [filepath]
  ;; Assume no header lines at top of input files
  (with-open [rdr (io/reader filepath)]
    (let [lines (line-seq rdr)
          separator (some-> (first lines)
                            (line->separator))]
      (when separator
        (->> lines
             (map (comp #(update % :DateOfBirth parse-datestring)
                        #(zipmap row-headers %)
                        #(map str/trim %)
                        #(str/split % separator)))
             doall))))) ;; Force in-memory realization now before file closes

(def env-spec [:map [:filepaths [:string {:min 1}]
                     :column-length [:string {:optional true
                                              :min 1}]]])

(defn validate-env [] (m/validate env-spec env))
(defn explain-env [] (m/explain env-spec env))

(defn string-compare-lowercase-asc [a b] (compare (str/lower-case a) (str/lower-case b)))
(defn string-compare-lowercase-desc [a b] (compare (str/lower-case b) (str/lower-case a)))

(def color-lastname [[:FavoriteColor string-compare-lowercase-asc]
                     [:LastName string-compare-lowercase-asc]])

(def dob [[:DateOfBirth compare]])

(def lastname-desc [[:LastName string-compare-lowercase-desc]])

(defn sort-records
  "Returns sorted list of input records based on sort-definition.
  records - seq of maps representing parsed records
  sort-definition - vec of vecs, where each 2-element subvec is keyfn & comparator"
  [records sort-definition]
  (let [sort-keys (map first sort-definition)
        keyfn (apply juxt sort-keys)
        sort-comparators (map second sort-definition)
        comparator (fn [a b]
                     (loop [[comparator & comp-rest] sort-comparators
                            [a1 & a-rest] a
                            [b1 & b-rest] b]
                       (let [result (comparator a b)]
                         (if (and (zero? result)
                                  (seq sort-comparators))
                           (recur comp-rest a-rest b-rest)
                           result))))]
    (sort-by keyfn comparator records)))

(defn run
  [{:keys [filepaths]}]
  (as-> filepaths <>
    (str/split filepaths #",")
    (map load-record <>)
    (flatten <>)))

(def record-spec [:map [:LastName [:string {:min 1}]
                        :FirstName [:string {:min 1}]
                        :Email [:string {:min 5}]
                        :FavoriteColor [:string {:min 4 :max 40}]
                        :DateOfBirth inst?]])

;; Register `run` with Malli. This definition reads "takes env-spec & returns record"
(m/=> run
      [:=> [:cat env-spec]
       [:sequential record-spec]])

(def output-views [color-lastname
                   dob
                   lastname-desc])

(defn trunc
  "This is unexpectedly missing in clojure.string"
  [s n]
  (subs s 0 (min (count s) n)))

(defn force-str-to-length
  "Forces string to specific width, padding with empty spaces as necessary.
  Longer strings are truncated with trailing ..."
  [length s]
  (let [difference (- length (count s))]
    (cond
      (pos? difference) (apply str s (repeat difference " "))
      (neg? difference) (str (trunc s (- length 3)) "...")
      :else s)))

(defn render-view-cast-to-string
  "Casts any record value to string"
  [x]
  (cond
    (string? x) x
    (inst? x) (.format (SimpleDateFormat. "M/d/yyyy") x)
    :else (throw (Exception. (str "Invalid value <x,typeOf(x)> " x (type x))))))

;; Little kludge to derive this from env. In a larger project,
;; we'd parse env into a final map, validate it, & expose it via state management lib
(defn env-column-length
  []
  (try
    (if (integer? (:column-length environ.core/env))
      (:column-length environ.core/env)
      (Integer/parseInt (:column-length environ.core/env)))
    (catch Exception _
      16)))

(defn render-view-data
  "Given sequence of maps with common keys, renders into ASCII table of minimum width"
  ([mapseq]
   (render-view-data mapseq nil))
  ([mapseq {:keys [column-length]
            :or {column-length (env-column-length)}}]
   (let [map-keys (keys (first mapseq))
         labels (->> (map name map-keys)
                     (map #(-> (force-str-to-length column-length %)
                               (str " ")))
                     (str/join)
                     (drop-last)
                     (str/join ""))
         lines (->> (map vals mapseq)
                    (map #(map (fn [s]
                                 (str (->> (render-view-cast-to-string s)
                                           (force-str-to-length column-length))
                                      " ")) %))
                    (map (comp str/join drop-last str/join))
                    (str/join "\n"))]
     (str labels "\n"
          (str/join (repeat (count labels) "="))
          "\n"
          lines))))

(defn run-render-views
  [env]
  (as-> (run (merge environ.core/env env)) <>
    (repeat <>)
    (map sort-records <> output-views)
    (map render-view-data <>)))

(defn -main
  [& _]
  ;; You could accept a file via Java program ARGS... but for sake of simplicity, just use $env
  (if-not (validate-env)
    (->> (explain-env)
         (me/humanize)
         (prn-str)
         (str "[ERROR] Problems below with input $env variables. Please fix & retry\n")
         (println))
    (->> (run-render-views env)
         (str/join "\n\n")
         (println))))

;; This technically belongs in user.clj or test, but including here to make usage obvious
(defmacro with-env
  [env & body]
  `(with-redefs [environ.core/env ~env]
     ~@body))

(comment ;; Try at a REPL. Modify to input different files
  (with-env {:filepaths "data/basic_record.csv,data/basic_record.psv,data/basic_record.ssv"
             :column-length 12}
    (-main)))
